### PR TITLE
NetSim: Interrouter messaging

### PR DESCRIPTION
**Pull request by islemaster**
Now that multipart addresses are supported, this PR introduces the next step of allowing messages to be sent between routers on the shard.
# What it does

If a router receives a message and can't find the `toAddress` in its local network, it will look to see if any other nodes on the shard have the given address.  If they do, the message gets forwarded to that node's router, which will then pass the message to the destination node.

Here it is in action!

send_between_routers
# What it doesn't do

For now, we assume that every router has a direct connection to every other router.  We aren't modeling any wires between routers.  Thus, an interrouter message will always travel

```
ClientA --> RouterA --> RouterB --> ClientB
```

and will leave two router log entries (on RouterA and RouterB).

There's no support yet for different paths to the destination or out-of-order routing - that's on the way.

DNS features have not yet been extended to the full network - the DNS tab will not show hosts/addresses around other routers, and the auto-DNS can't answer questions about nodes not in its subnet.
# Enabling the feature in levelbuilder

Sending messages between routers is only possible when the "Connected routers" option is checked in levelbuilder.  It's already been [documented on the wiki](http://wiki.code.org/display/PROD/Internet+Simulator+Level+Configuration+Options#InternetSimulatorLevelConfigurationOptions-SimulationConfiguration).
# How it works

The local client now acts as a proxy for each router along the outgoing message's path, up until its last router -> client leg where it passes responsibility to the receiving client.

```
Client A --> Router A --> Router B --> Client B
FROM     | TO       | SIMULATED BY
Client A | Router A | Client A
Router A | Router B | Client A
Router B | Client B | Client B
```

This requires every client to run a local simulation of every router on their shard; fortunately, most of those router simulations will do very little.

I'm delegating responsibility this way in anticipation of the next case (which we'll hit soon), which uses a router that's not directly 'owned' by either client:

```
Client A --> Router A --> Router C --> Router B --> Client B
FROM     | TO       | SIMULATED BY
Client A | Router A | Client A
Router A | Router C | Client A
Router C | Router B | Client A
Router B | Client B | Client B
```

### CODE REVIEW

File: apps/src/netsim/NetSimRouterNode.js [hunk 1]
```diff
@@ -314,11 +322,7 @@ NetSimRouterNode.create = function (shard, onComplete) {
       router.heartbeat = heartbeat;
       router.heartbeat.setBeatInterval(12000);
 
-      // Always try and update router immediately, to set its DisplayName
-      // correctly.
-      router.update(function (err) {
-        onComplete(err, router);
-      });
+      onComplete(null, router);
```
Comment (islemaster) [inline apps/src/netsim/NetSimRouterNode.js]: Unrelated: This is an artifact of an old setup where router displaynames were not calculated locally from their node ID.

File: apps/src/netsim/NetSimRouterNode.js [hunk 2]
```diff
@@ -759,22 +768,18 @@ NetSimRouterNode.prototype.initializeSimulation = function (nodeID) {
     var nodeChangeEvent = this.shard_.nodeTable.tableChange;
     var nodeChangeHandler = this.onNodeTableChange_.bind(this);
     this.nodeChangeKey_ = nodeChangeEvent.register(nodeChangeHandler);
-    logger.info("Router registered for nodeTable tableChange");
     
     var wireChangeEvent = this.shard_.wireTable.tableChange;
     var wireChangeHandler = this.onWireTableChange_.bind(this);
     this.wireChangeKey_ = wireChangeEvent.register(wireChangeHandler);
-    logger.info("Router registered for wireTable tableChange");
```
Comment (islemaster) [inline apps/src/netsim/NetSimRouterNode.js]: These log messages stopped being useful.

File: apps/src/netsim/NetSimRouterNode.js [hunk 3]
```diff
@@ -1463,17 +1530,16 @@ NetSimRouterNode.prototype.forwardMessageToRecipient_ = function (message, onCom
...
+    this.log(message.payload, NetSimLogEntry.LogStatus.DROPPED);
     onComplete(null);
     return;
-  }
-
-  var destinationNodeID = this.getNodeIDForAddress_(toAddress);
-  if (destinationNodeID === undefined) {
-    logger.warn("Destination address not in local network");
+  } else if (destinationNode === this && toAddress === this.getAddress()) {
+    // This router IS the packet's destination, it's done.
```
Comment (islemaster) [inline apps/src/netsim/NetSimRouterNode.js]: The `toAddress === this.getAddress()` check works around the way the Auto-DNS runs on the router node but has its own address.

File: apps/src/netsim/NetSimClientNode.js [hunk 4]
```diff
@@ -45,43 +46,66 @@ NetSimClientNode.prototype.getNodeType = function () {
...
+  if (!wire) {
+    return undefined;
   }
+  return wire.localAddress;
+};
 
-  return i18n.notConnected();
+/**
+ * Based on cached wire data, retrieve this node's outgoing wire.
+ * @returns {NetSimWire} null if wire does not exist.
```
Comment (bcjordan) [inline apps/src/netsim/NetSimClientNode.js]: jsdoc `|null`?

**Comment by bcjordan** [comment]
LGTM. Nice testing and writeup!

And this is so cool!

File: apps/src/netsim/NetSimClientNode.js (hunk 4 above)
Comment (islemaster) [inline apps/src/netsim/NetSimClientNode.js]: I think I like `|null` better.  I stumbled on this a little; if `{!NetSimWire}` is non-nullable and `{?NetSimWire}` is nullable, is `{NetSimWire}` just invalid (or under-informative) JSDoc?

I think this just strikes me as a bit of overspec-ing by JSDoc that I dislike.  Either the `?` or `!` modifier should exist, but not both.

**Pull request merged**